pkg/bucket/bandwidth/reader.go

```diff
@@ -0,0 +1,86 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package bandwidth
+
+import (
+	"io"
+	"time"
+)
+
+// MonitoredReader monitors the bandwidth
+type MonitoredReader struct {
+	bucketToken int
+	objectToken int
+	outgoing    chan<- Update
+	reader      io.Reader
+	startTime   time.Time
+	lastStop    time.Time
+	count       uint64
+	headerSize  int
+}
+
+// NewMonitoredReader returns a io.ReadCloser that reports bandwidth details
+func NewMonitoredReader(monitor *Monitor, bucket string, object string, reader io.Reader, headerSize int) *MonitoredReader {
+	timeNow := time.Now()
+	b, o := monitor.Track(bucket, object)
+	return &MonitoredReader{
+		bucketToken: b,
+		objectToken: o,
+		reader:      reader,
+		outgoing:    monitor.Incoming,
+		startTime:   timeNow,
+		lastStop:    timeNow,
+		headerSize:  headerSize,
+	}
+}
+
+// Read wraps the read reader
+func (m *MonitoredReader) Read(p []byte) (n int, err error) {
+	n, err = m.reader.Read(p)
+	if err != nil && err != io.EOF {
+		return
+	}
+	stop := time.Now()
+	if m.count == 0 {
+		m.count += uint64(m.headerSize)
+		m.headerSize = 0
+	}
+	m.count += uint64(n)
+	m.outgoing <- Update{
```

Comment:
Instead of dealing with channel mechanics, just have it send it directly with a function call. If the parent tracker is reasonably fast it shouldn't be a problem.